File: JAVA_ALGO/sort/QuickSort.java
import java.io.InputStreamReader;
import java.nio.Buffer;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.Arrays;

public class QuickSort {
    static void swap(int[] a, int idx1, int idx2){
        int temp = a[idx1];
        a[idx1] = a[idx2];
        a[idx2] = temp;
    }

    static void quickSort(int[] a, int left, int right){
        int pl = left;
        int pr = right;
        int x = a[(pl+pr)/2];

        do{
            while(a[pl] < x)    pl++;
            while(a[pr] > x)    pr--;
            if(pr>=pl)  swap(a, pl++, pr--);    // 양끝에서 한칸씩 검사하다가, 중간값보다 큰 pl, 중간값보다 작은 pr을 찾으면, 그 위치를 바꿔준다 -> 중간값 기준으로 작은게 왼쪽, 큰게 오른쪽으로 오게
        }while(pr>=pl);
        // 주어진 배열 a의 중간값 기준으로 작은게 왼쪽, 큰게 오른쪽으로 정리가 다 되면 pr은 왼쪽 구역의 제일 오른쪽값을, pl은 오른쪽 구역의 제일 왼쪽값을 가르킨다
        if(left<pr)    quickSort(a, left, pr);  // 정리된 배열의 왼쪽 구역을 한번 더 quicksort
        if(right>pl)   quickSort(a, pl, right); // 정리된 배열의 오른쪽 구역을 한번 더 quick sort
    }

    // non-recursive method
    static void quickSort2(int[] a, int left, int right){
        intStack lstack = new intStack(right - left +1);
        intStack rstack = new intStack(right -left+1);

        lstack.push(left);
        rstack.push(right);

        while(lstack.isEmpty() != true){
            int pl = left = lstack.pop();
            int pr = right = rstack.pop();
            int x = a[(pl+pr)/2];
            
            do{
                while(a[pl]<x)  pl++;
                while(a[pr]>x)  pr--;
                if(pl<=pr)  swap(a, pl++, pr--);
            }while(pr>=pl);

            if(left<pr){
                lstack.push(left);
                rstack.push(pr);
            }
            if(pl<right){
                lstack.push(pl);
                rstack.push(right);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        System.out.println("Quick Sort");
        System.out.print("number of array : ");

        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        int nx = Integer.parseInt(bf.readLine());
        int[] x = new int[nx];
        
        for(int i=0;i<nx;i++){
            System.out.print("x["+i+"] : ");
            x[i] = Integer.parseInt(bf.readLine());
        }

        System.out.println(Arrays.toString(x)); // input된 배열의 data 확인시, 정상

        quickSort2(x, 0, nx-1);

        System.out.println("upstreaming");
        for(int i=0;i<nx;i++)   System.out.print("x["+i+"] : "+x[i]+" ");
    }
}

File: PYTHON_ALGO/SORTING/quick_sort.py
# quick sort 
# 임의의 element 에다가 pivot 잡고          >> 맨왼쪽, 맨오른쪽 element 로 pivot 을 잡는 경우가 많은데, 계산시간이 길어질 수도 있다
# 범위 - 기준 - 비교 - swap                >> random element 로 pivot 을 잡기도 한다
# pivot 을 대상으로 비교할 element 의 범위를 설정하고
# 작은거 왼쪽 큰거 오른쪽 (예시)
# 작은거와 큰거로 나눈 가운데로 pivot 을 옮긴다
# time complexity
# worst O(n^2)                          >> 이미 정렬된 경우, 남은 부분을 다 훑어야 하므로
# normal & best O(nlog(n))

# pivot 기준으로 작은거 왼쪽, 큰거 오른쪽으로 나누는 함수
def partition(arr, left, right):
    # 맨 오른쪽 값을 피봇으로 잡고
    pivot = arr[right]
    # 값을 비교할 커서를 정의
    i = left - 1
    # 맨 왼쪽부터 피봇 직전까지 훑으면서
    for j in range(left, right):
        # 피봇보다 작은 값이면
        if arr[j] <= pivot:
            # 커서를 하나 증가시키고
            i += 1
            # 옮겨진 커서의 위치와, 값을 비교한 j끼리 swap
            arr[i], arr[j] = arr[j], arr[i]
    # 커서 i 를 기준으로 왼쪽에는 pivot 보다 작은 값, 오른쪽은 pivot 보다 큰 값
    # i 바로 옆으로 pivot으로 잡았던 arr element 이동
    arr[i+1], arr[right] = arr[right], arr[i+1]
    # pivot 의 위치 출력
    return i+1

# quick sort, recursive 구현
def quickSort(arr, left, right):
    # 왼쪽 값이 오른쪽 값보다 클때만 진행, 재귀함수 탈출을 위한 조건
    if left < right:
        # 위 partition 과정을 통해 구해진 partition 의 위치를 기준으로
        pivot = partition(arr, left, right)
        # 왼쪽 정렬 (pivot 보다 작았던 값들)
        quickSort(arr, left, pivot-1)
        # 오른쪽 정렬 (pivot 보다 컷던 값들)
        quickSort(arr, pivot+1, right)
    return arr


data = [10, 55, 23, 2, 79, 101, 16, 82, 30, 45]
arr = [6, 3, 2, 10, 10, 10,-10,-10, 7, 3]
print(quickSort(data, 0, len(data)-1))
print(quickSort(arr, 0, len(arr)-1))